In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [3]:
df=pd.read_excel(r"D:\DATA SCIENCE\srikanth bi project\bi\vehicles_for_sale.xlsx")
pd.options.display.max_columns = None
df

Make  Year Condition   Mileage   Price  Seller Rating  \
0       Lexus  2024       New       0.0  112865            4.7   
1       Acura  2007      Used   61110.0   11295            4.2   
2     McLaren  2016      Used    6305.0  219997            3.1   
3        Audi  2016      Used   65715.0   23999            3.6   
4         BMW  2018      Used   19830.0   39799            4.4   
...       ...   ...       ...       ...     ...            ...   
9011     Jeep  2024       New      10.0   52301            4.0   
9012      BMW  2022      Used   48804.0   27979            4.8   
9013      GMC  2024       New       3.0   51080            4.5   
9014      GMC  2012      Used  146694.0    8995            NaN   
9015   Subaru  2024       New       4.0   42014            NaN   

      Seller Rating Count   Interior color          Drivetrain  Min MPG  \
0                  1261.0           Black    Rear-wheel Drive      15.0   
1                   440.0             –   Front-wheel Drive      22.0   
2                   421.0    Carbon Black    Rear-wheel Drive       NaN   
3                   123.0           Black     All-wheel Drive      23.0   
4                    91.0           Black     All-wheel Drive       0.0   
...                   ...              ...                 ...      ...   
9011                 54.0    Global Black    Four-wheel Drive      18.0   
9012               3739.0  Canberra Beige     All-wheel Drive      21.0   
9013               3896.0           Black    Four-wheel Drive      17.0   
9014                  NaN       Jet Black     All-wheel Drive      20.0   
9015                  NaN           Black     All-wheel Drive      22.0   

      Max MPG  Fuel type  Engine Size (L)  Accidents or damage  \
0        25.0  Gasoline               5.0                  NaN   
1        31.0  Gasoline               2.4                  1.0   
2         NaN  Gasoline               3.8                  1.0   
3        27.0  Gasoline               2.0                  1.0   
4         0.0  Gasoline               2.0                  0.0   
...       ...        ...              ...                  ...   
9011     25.0  Gasoline               3.6                  NaN   
9012     28.0  Gasoline               2.0                  0.0   
9013     20.0  Gasoline               5.3                  NaN   
9014     22.0  Gasoline               2.4                  1.0   
9015     29.0  Gasoline               2.4                  NaN   

      1-owner vehicle  Personal use only  Clean title  Open recall  \
0                 NaN                NaN          NaN            0   
1                 1.0                1.0          NaN            0   
2                 0.0                0.0          NaN            0   
3                 0.0                1.0          1.0            0   
4                 1.0                0.0          1.0            0   
...               ...                ...          ...          ...   
9011              NaN                NaN          NaN            0   
9012              1.0                0.0          NaN            0   
9013              NaN                NaN          NaN            0   
9014              0.0                1.0          NaN            1   
9015              NaN                NaN          NaN            0   

      Comfort_score  Interior_score  Performance_score  Value_score  \
0               NaN             NaN                NaN          NaN   
1               NaN             NaN                NaN          NaN   
2               5.0             5.0                5.0          5.0   
3               3.0             5.0                5.0          5.0   
4               5.0             5.0                5.0          5.0   
...             ...             ...                ...          ...   
9011            5.0             5.0                5.0          5.0   
9012            NaN             NaN                NaN          NaN   
9013            5.0             5.0       

In [4]:
df.shape

(9016, 24)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9016 entries, 0 to 9015
Data columns (total 24 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Make                 9016 non-null   object 
 1   Year                 9016 non-null   int64  
 2   Condition            9016 non-null   object 
 3   Mileage              8867 non-null   float64
 4   Price                9016 non-null   int64  
 5   Seller Rating        7661 non-null   float64
 6   Seller Rating Count  7661 non-null   float64
 7   Interior color       9016 non-null   object 
 8   Drivetrain           9016 non-null   object 
 9   Min MPG              7421 non-null   float64
 10  Max MPG              7421 non-null   float64
 11  Fuel type            8441 non-null   object 
 12  Engine Size (L)      8251 non-null   float64
 13  Accidents or damage  5400 non-null   float64
 14  1-owner vehicle      5374 non-null   float64
 15  Personal use only    5400 non-null   f

### we can observe that there are 5 object datatype features, 3 integer datatype features and 16 float datatype features

In [6]:
df.describe()

Year        Mileage         Price  Seller Rating  \
count  9016.000000    8867.000000  9.016000e+03    7661.000000   
mean   2019.456300   36301.668546  4.008421e+04       4.381504   
std       6.449194   44208.902388  4.259945e+04       0.609913   
min    1959.000000       0.000000  1.997000e+03       1.300000   
25%    2017.000000      10.000000  2.149000e+04       4.200000   
50%    2022.000000   18944.000000  3.238300e+04       4.600000   
75%    2024.000000   62501.000000  4.899100e+04       4.800000   
max    2025.000000  324349.000000  1.699800e+06       5.000000   

       Seller Rating Count      Min MPG      Max MPG  Engine Size (L)  \
count          7661.000000  7421.000000  7421.000000      8251.000000   
mean           1432.193578    20.775771    26.510982         2.964659   
std            1592.647184     7.217294     8.593017         1.325837   
min               5.000000     0.000000     0.000000         0.000000   
25%             235.000000    17.000000    24.000000         2.000000   
50%             990.000000    21.000000    28.000000         2.500000   
75%            2044.000000    25.000000    32.000000         3.600000   
max           11663.000000   150.000000   122.000000         8.400000   

       Accidents or damage  1-owner vehicle  Personal use only  Clean title  \
count          5400.000000      5374.000000        5400.000000  1331.000000   
mean              0.330000         0.498139           0.676296     0.995492   
std               0.470256         0.500043           0.467932     0.067014   
min               0.000000         0.000000           0.000000     0.000000   
25%               0.000000         0.000000           0.000000     1.000000   
50%               0.000000         0.000000           1.000000     1.000000   
75%               1.000000         1.000000           1.000000     1.000000   
max               1.000000         1.000000           1.000000     1.000000   

       Open recall  Comfort_score  Interior_score  Performance_score  \
count  9016.000000    4386.000000     4386.000000        4386.000000   
mean      0.071761       4.598495        4.563383           4.543776   
std       0.258107       0.838355        0.818943           0.865939   
min       0.000000       1.000000        1.000000           1.000000   
25%       0.000000       4.000000        4.000000           4.000000   
50%       0.000000       5.000000        5.000000           5.000000   
75%       0.000000       5.000000        5.000000           5.000000   
max       1.000000       5.000000        5.000000           5.000000   

       Value_score  Exterior_score  Reliability_score  
count  4386.000000     4386.000000        4310.000000  
mean      4.407889        4.727086           4.484455  
std       1.028365        0.697989           1.039983  
min       1.000000        1.000000           1.000000  
25%       4.000000        5.000000           4.000000  
50%       5.000000        5.000000           5.000000  
75%       5.000000        5.000000           5.000000  
max       5.000000        5.000000           5.000000

In [7]:
df.describe(include='object')

Make Condition Interior color        Drivetrain  Fuel type
count   9016      9016           9016              9016       8441
unique    57         3            611                10          8
top     Ford      Used         Black   All-wheel Drive   Gasoline 
freq     847      5005           3434              3911       7553

In [8]:
df.duplicated().sum()

2485

### we can observe that there are 2485 duplicate records that can be dropped

In [9]:
df.isnull().sum()/df.shape[0]*100

Make                    0.000000
Year                    0.000000
Condition               0.000000
Mileage                 1.652618
Price                   0.000000
Seller Rating          15.028838
Seller Rating Count    15.028838
Interior color          0.000000
Drivetrain              0.000000
Min MPG                17.690772
Max MPG                17.690772
Fuel type               6.377551
Engine Size (L)         8.484916
Accidents or damage    40.106477
1-owner vehicle        40.394854
Personal use only      40.106477
Clean title            85.237356
Open recall             0.000000
Comfort_score          51.353150
Interior_score         51.353150
Performance_score      51.353150
Value_score            51.353150
Exterior_score         51.353150
Reliability_score      52.196096
dtype: float64

In [10]:
(df.isnull().sum()/df.shape[0]*100).sort_values(ascending=False)

Clean title            85.237356
Reliability_score      52.196096
Exterior_score         51.353150
Value_score            51.353150
Performance_score      51.353150
Interior_score         51.353150
Comfort_score          51.353150
1-owner vehicle        40.394854
Personal use only      40.106477
Accidents or damage    40.106477
Min MPG                17.690772
Max MPG                17.690772
Seller Rating          15.028838
Seller Rating Count    15.028838
Engine Size (L)         8.484916
Fuel type               6.377551
Mileage                 1.652618
Year                    0.000000
Open recall             0.000000
Drivetrain              0.000000
Interior color          0.000000
Price                   0.000000
Condition               0.000000
Make                    0.000000
dtype: float64

# Preprocessing

### Dropping the duplicate records

In [47]:
df1=df.copy()

In [48]:
df1.duplicated(keep='first').sum()

2485

In [49]:
df1=df1.drop_duplicates()
df1.shape

(6531, 24)

### checking all unique values

In [50]:
for i in df1.columns:
    print(i,'\n',df1[i].unique(),'\n')

Make 
 ['Lexus' 'Acura' 'McLaren' 'Audi' 'BMW' 'Jeep' 'Cadillac' 'Nissan'
 'Chrysler' 'Ford' 'Toyota' 'INFINITI' 'Volkswagen' 'Chevrolet' 'Jaguar'
 'Subaru' 'Mercedes-Benz' 'Pontiac' 'Triumph' 'Scion' 'Porsche' 'Aston'
 'Mazda' 'Honda' 'Lincoln' 'Saturn' 'Lotus' 'Ferrari' 'Rolls-Royce'
 'Genesis' 'Dodge' 'Hyundai' 'GMC' 'Mitsubishi' 'Kia' 'Maserati'
 'Lamborghini' 'Rivian' 'Buick' 'RAM' 'Volvo' 'Polestar' 'Hummer'
 'Delorean' 'Mercury' 'Saab' 'FIAT' 'Bentley' 'Fisker' 'MINI' 'Isuzu'
 'Land' 'Alfa' 'Oldsmobile' 'Bugatti' 'Tesla' 'Plymouth'] 

Year 
 [2024 2007 2016 2018 2019 2012 2008 2013 2002 2020 2014 2015 2004 2000
 2011 1988 2023 2006 1975 2009 2005 2017 2010 2022 2021 2003 1983 1981
 2001 1993 1971 1985 1980 1974 1999 1989 1972 1994 1992 1991 1984 1976
 1995 2025 1998 1997 1990 1970 1996 1982 1960 1959 1973 1987 1969 1966] 

Condition 
 ['New' 'Used' 'Certified'] 

Mileage 
 [    0. 61110.  6305. ...  3246. 40291. 79767.] 

Price 
 [112865  11295 219997 ...  41795  30952  51080] 


In [51]:
for i in df1.columns:
    print(i,'\n',df1[i].value_counts(),'\n')

Make 
 Make
Ford             636
Jeep             596
Chevrolet        595
Nissan           428
Mercedes-Benz    332
BMW              326
Honda            285
Subaru           278
Toyota           264
Audi             226
Cadillac         219
Kia              214
Lexus            208
Volkswagen       206
RAM              171
INFINITI         136
Hyundai          135
Dodge            113
Tesla            111
Lincoln          102
GMC               88
Mazda             87
Volvo             80
Chrysler          77
Buick             72
Porsche           68
Acura             64
Jaguar            42
Land              41
Mitsubishi        39
Maserati          31
Ferrari           29
Genesis           28
MINI              27
Alfa              25
Rivian            19
Rolls-Royce       18
Scion             18
Aston             16
Lamborghini       15
Pontiac           11
Bentley            9
Saab               7
Polestar           6
McLaren            6
Hummer             5
Saturn             5
M

In [52]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6531 entries, 0 to 9013
Data columns (total 24 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Make                 6531 non-null   object 
 1   Year                 6531 non-null   int64  
 2   Condition            6531 non-null   object 
 3   Mileage              6431 non-null   float64
 4   Price                6531 non-null   int64  
 5   Seller Rating        5585 non-null   float64
 6   Seller Rating Count  5585 non-null   float64
 7   Interior color       6531 non-null   object 
 8   Drivetrain           6531 non-null   object 
 9   Min MPG              5357 non-null   float64
 10  Max MPG              5357 non-null   float64
 11  Fuel type            6110 non-null   object 
 12  Engine Size (L)      5974 non-null   float64
 13  Accidents or damage  3897 non-null   float64
 14  1-owner vehicle      3878 non-null   float64
 15  Personal use only    3897 non-null   float6

### There is a incorrect value '– ' in several columns so lets replace them with mode in categorical variables and mean in numerical variables

In [53]:
for i in df1.columns:
    df1[i].replace({'– ':df1[i].mode()[0]},inplace=True)

### Lets also replace 'open recall' values 0 and 1 with 'recall' and 'no recall'

In [54]:
df1['Open recall'].replace([0,1],['No Recall','Recall'],inplace=True)

### Lets replace "0.0" in 'Min MPG','Max MPG','Engine Size (L)' with nan

In [55]:
for i in ['Min MPG','Max MPG','Engine Size (L)']:
    df1[i].replace([0.0],np.nan,inplace=True)

# Dropping the null values

In [56]:
df2=df1.copy()

In [57]:
(df2.isnull().sum()/df2.shape[0]*100).sort_values(ascending=False)

Clean title            85.607105
Reliability_score      52.243148
Exterior_score         51.523503
Value_score            51.523503
Performance_score      51.523503
Interior_score         51.523503
Comfort_score          51.523503
1-owner vehicle        40.621651
Personal use only      40.330730
Accidents or damage    40.330730
Max MPG                22.798959
Min MPG                21.313734
Seller Rating          14.484765
Seller Rating Count    14.484765
Engine Size (L)         8.620426
Fuel type               6.446180
Mileage                 1.531159
Year                    0.000000
Open recall             0.000000
Drivetrain              0.000000
Interior color          0.000000
Price                   0.000000
Condition               0.000000
Make                    0.000000
dtype: float64

### 'Clean title' has 85% null records and i think it doesnt effect the price rate so i will drop this column

In [58]:
df2.drop('Clean title',axis=1,inplace=True)

In [59]:
(df2.isnull().sum()/df2.shape[0]*100).sort_values(ascending=False)

Reliability_score      52.243148
Exterior_score         51.523503
Value_score            51.523503
Performance_score      51.523503
Interior_score         51.523503
Comfort_score          51.523503
1-owner vehicle        40.621651
Personal use only      40.330730
Accidents or damage    40.330730
Max MPG                22.798959
Min MPG                21.313734
Seller Rating          14.484765
Seller Rating Count    14.484765
Engine Size (L)         8.620426
Fuel type               6.446180
Mileage                 1.531159
Year                    0.000000
Open recall             0.000000
Drivetrain              0.000000
Interior color          0.000000
Price                   0.000000
Condition               0.000000
Make                    0.000000
dtype: float64

### 'Reliability_score' has 52% null values and 'Interior_score', 'Performance_score', 'Value_score', 'Exterior_score','Reliability_score' have 51% null values so lets fill it with "un reported"

In [60]:
for i in ['Comfort_score','Interior_score', 'Performance_score', 'Value_score', 'Exterior_score','Reliability_score']:
    df2[i].replace([np.nan,1.0,2.0,3.0,4.0,5.0],["Not Reported",1,2,3,4,5],inplace=True)

In [61]:
(df2.isnull().sum()/df2.shape[0]*100).sort_values(ascending=False)

1-owner vehicle        40.621651
Accidents or damage    40.330730
Personal use only      40.330730
Max MPG                22.798959
Min MPG                21.313734
Seller Rating          14.484765
Seller Rating Count    14.484765
Engine Size (L)         8.620426
Fuel type               6.446180
Mileage                 1.531159
Open recall             0.000000
Exterior_score          0.000000
Value_score             0.000000
Performance_score       0.000000
Interior_score          0.000000
Comfort_score           0.000000
Make                    0.000000
Year                    0.000000
Drivetrain              0.000000
Interior color          0.000000
Price                   0.000000
Condition               0.000000
Reliability_score       0.000000
dtype: float64

### In the column 'Accidents or damage' there are 40% null values, lets replace null values with 'not reported', 1.0 with 'Accident' and 0.0 with 'No Accident'

In [62]:
df2['Accidents or damage'].replace([np.nan,1.0,0.0],["Not Reported","Accident", "No Accident"],inplace=True)

### In the column '1-owner vehicle' there are 40% null values, lets replace null values with 'not reported', 1.0 with "Single owner" and 0.0 with "Multiple Owners"

In [63]:
df2['1-owner vehicle'].replace([np.nan,1.0,0.0],["Not Reported","Single owner", "Multiple Owners"],inplace=True)

### In the column 'Personal use only' there are 40% null values, lets replace null values with 'not reported', 1.0 with "Personal use" and 0.0 with "Not Personal use"

In [64]:
df2['Personal use only'].replace([np.nan,1.0,0.0],["Not Reported","Personal use", "Not Personal use"],inplace=True)

In [65]:
(df2.isnull().sum()/df2.shape[0]*100).sort_values(ascending=False)

Max MPG                22.798959
Min MPG                21.313734
Seller Rating          14.484765
Seller Rating Count    14.484765
Engine Size (L)         8.620426
Fuel type               6.446180
Mileage                 1.531159
Personal use only       0.000000
Exterior_score          0.000000
Value_score             0.000000
Performance_score       0.000000
Interior_score          0.000000
Comfort_score           0.000000
Open recall             0.000000
Make                    0.000000
1-owner vehicle         0.000000
Accidents or damage     0.000000
Year                    0.000000
Drivetrain              0.000000
Interior color          0.000000
Price                   0.000000
Condition               0.000000
Reliability_score       0.000000
dtype: float64

### Lets fill the null values above 10% with mean in the numerical columns and mode in the categorical columns

In [66]:
for i in ['Min MPG','Max MPG','Seller Rating Count']:
    if df2[i].dtype in ['int64','float64']:
        df2[i].fillna(df2[i].mean().round(0),inplace=True)
    else:
        df2[i].fillna(df2[i].mode()[0],inplace=True)

In [67]:
df2['Seller Rating'].fillna(df2['Seller Rating'].mean().round(1),inplace=True)

In [68]:
(df2.isnull().sum()/df2.shape[0]*100).sort_values(ascending=False)

Engine Size (L)        8.620426
Fuel type              6.446180
Mileage                1.531159
Exterior_score         0.000000
Value_score            0.000000
Performance_score      0.000000
Interior_score         0.000000
Comfort_score          0.000000
Open recall            0.000000
Personal use only      0.000000
1-owner vehicle        0.000000
Accidents or damage    0.000000
Make                   0.000000
Year                   0.000000
Max MPG                0.000000
Min MPG                0.000000
Drivetrain             0.000000
Interior color         0.000000
Seller Rating Count    0.000000
Seller Rating          0.000000
Price                  0.000000
Condition              0.000000
Reliability_score      0.000000
dtype: float64

### Because theres only less than 10% null values lets drop the null records

In [69]:
df2.dropna(inplace=True)

In [70]:
(df2.isnull().sum()/df2.shape[0]*100).sort_values(ascending=False)

Make                   0.0
Engine Size (L)        0.0
Exterior_score         0.0
Value_score            0.0
Performance_score      0.0
Interior_score         0.0
Comfort_score          0.0
Open recall            0.0
Personal use only      0.0
1-owner vehicle        0.0
Accidents or damage    0.0
Fuel type              0.0
Year                   0.0
Max MPG                0.0
Min MPG                0.0
Drivetrain             0.0
Interior color         0.0
Seller Rating Count    0.0
Seller Rating          0.0
Price                  0.0
Mileage                0.0
Condition              0.0
Reliability_score      0.0
dtype: float64

### Lets Rename 'Miles' to 'Miles Travelled'

In [71]:
df2.rename({'Mileage':"Miles Traveled"},axis=1,inplace=True)

### Lets reset the index

In [72]:
df2.reset_index(drop=True, inplace=True)

In [73]:
df2

Make  Year Condition  Miles Traveled   Price  Seller Rating  \
0             Lexus  2024       New             0.0  112865            4.7   
1             Acura  2007      Used         61110.0   11295            4.2   
2           McLaren  2016      Used          6305.0  219997            3.1   
3              Audi  2016      Used         65715.0   23999            3.6   
4               BMW  2018      Used         19830.0   39799            4.4   
...             ...   ...       ...             ...     ...            ...   
5853           Alfa  2024       New          3246.0   45179            4.8   
5854           Jeep  2024       New            11.0   45735            4.2   
5855           Audi  2021      Used         40291.0   41795            2.9   
5856  Mercedes-Benz  2024       New            11.0   62405            4.6   
5857            GMC  2024       New             3.0   51080            4.5   

      Seller Rating Count Interior color          Drivetrain  Min MPG  \
0                  1261.0         Black    Rear-wheel Drive      15.0   
1                   440.0         Black   Front-wheel Drive      22.0   
2                   421.0  Carbon Black    Rear-wheel Drive      22.0   
3                   123.0         Black     All-wheel Drive      23.0   
4                    91.0         Black     All-wheel Drive      22.0   
...                   ...            ...                 ...      ...   
5853                105.0         Black     All-wheel Drive      22.0   
5854               1632.0          B6x7    Four-wheel Drive      18.0   
5855                 43.0         Black     All-wheel Drive      18.0   
5856               1245.0     AMG Black     All-wheel Drive      23.0   
5857               3896.0         Black    Four-wheel Drive      17.0   

      Max MPG  Fuel type  Engine Size (L) Accidents or damage  \
0        25.0  Gasoline               5.0        Not Reported   
1        31.0  Gasoline               2.4            Accident   
2        28.0  Gasoline               3.8            Accident   
3        27.0  Gasoline               2.0            Accident   
4        28.0  Gasoline               2.0         No Accident   
...       ...        ...              ...                 ...   
5853     28.0  Gasoline               1.3        Not Reported   
5854     25.0  Gasoline               3.6        Not Reported   
5855     24.0  Gasoline               3.0            Accident   
5856     33.0  Gasoline               2.0        Not Reported   
5857     20.0  Gasoline               5.3        Not Reported   

      1-owner vehicle Personal use only Open recall Comfort_score  \
0        Not Reported      Not Reported   No Recall  Not Reported   
1        Single owner      Personal use   No Recall  Not Reported   
2     Multiple Owners  Not Personal use   No Recall             5   
3     Multiple Owners      Personal use   No Recall             3   
4        Single owner  Not Personal use   No Recall             5   
...               ...               ...         ...           ...   
5853     Not Reported      Not Reported   No Recall             5   
5854     Not Reported      Not Reported   No Recall             5   
5855     Single owner      Personal use   No Recall  Not Reported   
5856     Not Reported      Not Reported   No Recall  Not Reported   
5857     Not Reported      Not Reported   No Recall             5   

     Interior_score Performance_score   Value_score Exterior_score  \
0      Not Reported      Not Reported  Not Reported   Not Reported   
1      Not Reported      Not Reported  Not Reported   Not Reported   
2                 5                 5             5              5   
3                 5                 5             5              5   
4                 5                 5             5              5   
...             ...               ...           ...            ...   
5853              5                 5             5              5   
5854              5            

In [74]:
df3=df2.copy()

In [75]:
df3.to_excel("cars_dataset_cleaned.xlsx")